mlrun/api/db/sqldb/db.py

```diff
@@ -837,96 +693,1093 @@ def list_artifacts(
+        # 1. tag - in this case _resolve_tag will find the relevant uids and will return a list
+        # 2. uid - in this case _resolve_tag won't find anything and simply return what was given to it, which actually
+        # represents the uid
+        if isinstance(ids, list) and ids:
+            query = query.filter(Artifact.id.in_(ids))
+            db_tag = tag
+        elif isinstance(ids, str) and ids:
+            query = query.filter(Artifact.uid == ids)
+        else:
+            # Select by last updated
+            max_updated = session.query(func.max(Artifact.updated)).filter(
+                Artifact.project == project, Artifact.key == key
+            )
+            query = query.filter(Artifact.updated.in_(max_updated))
+
+        art = query.one_or_none()
+        if not art:
+            artifact_uri = generate_artifact_uri(project, key, tag, iter)
+            raise mlrun.errors.MLRunNotFoundError(f"Artifact {artifact_uri} not found")
+
+        artifact_struct = art.struct
+        # We only set a tag in the object if the user asked specifically for this tag.
+        if db_tag:
+            self._set_tag_in_artifact_struct(artifact_struct, db_tag)
+        return artifact_struct
+
+    def read_artifact_v2(
+        self,
+        session,
+        key: str,
+        tag: str = None,
+        iter: int = None,
+        project: str = None,
+        producer_id: str = None,
+        uid: str = None,
+        raise_on_not_found: bool = True,
+    ):
+        query = self._query(session, ArtifactV2, key=key, project=project)
+
+        computed_tag = tag or "latest"
+        artifact_tag_uid = None
+        if tag and not uid:
+            artifact_tag_uid = self._resolve_class_tag_uid(
+                session, ArtifactV2, project, key, computed_tag
+            )
+            if not artifact_tag_uid:
+                artifact_uri = generate_artifact_uri(project, key, tag, iter)
+                raise mlrun.errors.MLRunNotFoundError(
+                    f"Artifact {artifact_uri} not found"
+                )
+            uid = artifact_tag_uid
+
+        if uid:
+            query = query.filter(ArtifactV2.uid == uid)
+        if producer_id:
+            query = query.filter(ArtifactV2.producer_id == producer_id)
+
+        # keep the query without the iteration filter for later error handling
+        query_without_iter = query
+        if iter:
+            query = query.filter(ArtifactV2.iteration == iter)
+
+        db_artifact = query.one_or_none()
+
+        if not db_artifact:
+            # if the artifact was not found and iter==0, we might be looking for a link artifact
+            # in this case, we need to look for the artifact with the best iteration
+            fail = True
+            if iter == 0:
+                query_without_iter = query_without_iter.filter(
+                    ArtifactV2.best_iteration
+                )
+                db_artifact = query_without_iter.all()
```

Comment:
I agree regarding `one_or_none()`.

I used the `fail` param to avoid duplicating the `raise NotFound` excepetion.
if we entered the first `if not db_artifact:` condition, we only check again for the best iteration if `iter == 0`, but if it doesn't we want to just fail immediately.

if indeed `iter == 0`, then we will fail only if the second query returns `None`.

Reply:
👍